cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -738,6 +761,109 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriReal(const Uri& uri,
 
 }  // namespace
 
+Status RegisterFileSystemFactory(std::vector<std::string> schemes,
+                                 FileSystem::Factory factory) {
+  auto& [mutex, scheme_to_factory] = FileSystemFactoryRegistry();
+  std::unique_lock lock{mutex};
+
+  for (auto&& scheme : schemes) {
+    if (auto& ref = scheme_to_factory[scheme]) {
+      return Status::KeyError(
+          "Tried to add a factory for ", scheme,
+          ":// uris, but a factory was already registered for that scheme");
+    } else {
+      ref = factory;
+    }
+  }
+  return Status::OK();
+}
+
+// XXX alternative approach:
+// If we distrust global constructors more than we distrust
+// symbol visibility, we could export specially named functions
+// and retrieve them via dlsym.
+Status RegisterFileSystemFactoryModule(std::string module_name) {
+  static auto* handle = dlopen(nullptr, RTLD_NOW | RTLD_LOCAL);
```

Comment:
Is this changed to `dlopen(std::string("libarrow_fs_") + module_name + ".so", ...)` or something eventually, right?

Reply:
It could be, but I'd prefer to be less prescriptive about which library a module must reside in. In the example of `ArrowFileSystemModule_slow`, the module is actually in libarrow.so rather than libarrow_fs_slow.so. That flexibility seems useful to me